gtrsconvert/convert.go

```diff
@@ -0,0 +1,145 @@
+package gtrsconvert
+
+import (
+	"encoding"
+	"reflect"
+	"strconv"
+	"time"
+)
+
+// ConvertibleTo is implemented by types that can convert themselves to a map.
+type ConvertibleTo interface {
+	ToMap() (map[string]any, error)
+}
+
+// ConvertibleFrom is implemented by types that can load themselves from a map.
+type ConvertibleFrom interface {
+	FromMap(map[string]any) error
+}
+
+// structToMap convert a struct to a map.
+func StructToMap(st any) (map[string]any, error) {
+	if c, ok := st.(ConvertibleTo); ok {
+		return c.ToMap()
+	}
+
+	rv := reflect.ValueOf(st)
+	rt := reflect.TypeOf(st)
+	out := make(map[string]interface{}, rv.NumField())
+
+	for i := 0; i < rv.NumField(); i++ {
+		fieldValue := rv.Field(i)
+		fieldType := rt.Field(i)
+		fieldName := getFieldNameFromType(fieldType)
+		switch v := fieldValue.Interface().(type) {
+		case time.Time:
+			out[fieldName] = v.Format(time.RFC3339Nano)
+		case time.Duration:
+			out[fieldName] = v.String()
+		case encoding.TextMarshaler:
+			txt, err := v.MarshalText()
+			if err != nil {
+				return nil, SerializeError{
+					Field: fieldType.Name,
+					Value: fieldValue.Interface(),
+					Err:   err,
+				}
+			}
+			out[fieldName] = string(txt)
```

Comment:
Based on what I found in the `go-redis` library it uses only `Binary[Un]Marshaller` interfaces. On one side we'd allow text types to be send, on the other if a struct implements both Text/Binary, we'd prioritize text whereas the redis client would've used binary 🤔 Not sure... what do you think?

Reply:
didn't realize the credit client used binary - I think we should use the same.